### ISSUE TITLE
Builtin read not interruptible via signal

### ISSUE THREAD

**Issue by mil** [oilshell/oil#986]
I'm working on an interactive script with oil / osh that needs to be able to be interrupted while reading from user input. E.g. while the builtin `read` is being used.. there might be a SIGHUP sent to the script process (in which case I'd like to terminate the read).

```sh
#!/usr/bin/env osh
trap hup HUP

hup() {
  echo "sighup recvd"
}

echo foo
read bar
echo biz $bar
```

Behavior differs between osh, bash, and ash as:
**In osh:** Issuing `kill -HUP PID` will have osh wait until `read` completes, to run the function body of `hup`.
**In bash:** Issuing `kill -HUP PID` will have bash immediately run the function body of `hup` but not cancel the `read`.
**In ash:** Issuing `kill -HUP PID` will have ash immediately complete the `read` and then run the function body of `hup`.

Of all three, osh is the least preferable for my usecase as I can't shortcircut read at all.
Is there a way to have `read` interruptible via signals -- or is this a bug with `read`?

Tested against oil 0.9.0.

**Comment by andychu** [comment]
Thanks for the report!  This might be similar to #858 about `wait`, which I fixed

**Comment by andychu** [comment]
Note: may be related to

https://github.com/oilshell/oil/commit/04fc3c95ed7ffb4351ae92f07ecb6f73f2499597

which fixed a crash.  Maybe we should not retry on EINTR in C ... hm

**Comment by andychu** [comment]
notes: I grepped for `posix.read()`, which would be the function that should change

And it is used in several places in `osh/builtin_misc.py` for various kinds of `read`, and also in `core/executor.py` for command subs

**Comment by mil** [comment]
Yeah I would think `read` retry on `EINTR` is incorrect at:
https://github.com/oilshell/oil/blob/master/native/posixmodule.c#L1720

If I understand the reason for 04fc3c9; it was that it resolved #292 which caused a crash on `SIGWINCH`. There's already a signal handler for `SIGTERM` and `SIGKILL` in oil's binary I assume. As an alternative to re-reading on EINTR, why not just add a signal handler for `SIGWINCH` to nop? I assume that's what other shells do given their observable behavior.

Does that sound like the correct solution? I can help try and implement if you give a bit of guidance.

**Comment by andychu** [comment]
Yes the unconditional EINTR retry isn't desirable, because it disagrees with other shells, and makes it harder to abort a program.

I think I was waiting to do something "principled" for the whole interpreter, but I think it makes sense to just fix this bug for now.  The signal handling situation is a complicated because of the way we reuse the Python interpreter (which is intended to go away in the future).

- SIGTERM: the Python interpreter already turns this into `KeyboardInterrupt`.  I would rather it not do this for consistency!  
- SIGKILL: I think this is the one that the app can't handle, by definition

So I think what we need to do is turn EINTR back into a "return value" and then change every caller in Oil to handle it in some way or another.  There are not that many callers, which is why I grepped for it.

Or we could keep it an excepption, like Python does.  In Python you get IOError or OSError with `e.errno == errno.EINTR`.

Probably the hold-up is that I wanted to make it more like a "return value" like in C, because that's more explicit and all the `try/except` makes the code awkward.  But this is a surface issue and not fundamental -- really what we care about is the behavior!

So I'm open to any ideas  / info... !  Sorry for the delay; I do want to get back to this

**Comment by mil** [comment]
Thanks for the extra info @andychu - I'll see if when i free up I can poke around the dev build. This would be my first change working with oil source, so it may be a bit to ambitious but i'll see. Changing calls for `posix.read` and handling with `try/except` seems doable if I can get up to speed with the tooling / how I would add tests etc.

In the meantime I did manage a workaround via `wait` like:
```sh
#!/usr/bin/env osh

trap handler HUP

handler() {
  echo works
}

echo before
read var & wait
echo after
```

Sending `pkill -HUP osh` interrupts the read as desired this way.. the only unfortunate bit with this workaround is after quitting the script, to STDERR, a really ugly message get displayed:

```sh
Traceback (most recent call last):
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 328, in _cpython_main_hook
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 301, in main
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 292, in AppBundleMain
  File "/home/andy/git/oilshell/oil/core/shell.py", line 675, in Main
  File "/home/andy/git/oilshell/oil/core/main_loop.py", line 330, in Batch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1495, in ExecuteAndCatch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1342, in _Execute
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 988, in _Dispatch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1398, in _ExecuteList
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1342, in _Execute
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 652, in _Dispatch
  File "/home/andy/git/oilshell/oil/core/executor.py", line 295, in RunBackgroundJob
  File "/home/andy/git/oilshell/oil/core/process.py", line 880, in Start
  File "/home/andy/git/oilshell/oil/core/process.py", line 704, in Run
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1495, in ExecuteAndCatch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1342, in _Execute
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 636, in _Dispatch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 477, in _RunSimpleCommand
  File "/home/andy/git/oilshell/oil/core/executor.py", line 230, in RunSimpleCommand
  File "/home/andy/git/oilshell/oil/core/executor.py", line 126, in RunBuiltin
  File "/home/andy/git/oilshell/oil/osh/builtin_misc.py", line 328, in Run
  File "/home/andy/git/oilshell/oil/core/pyos.py", line 119, in Restore
termios.error: (5, 'I/O error')
FATAL: couldn't import from app bundle '/usr/bin/oil.ovm' (1)
Stripping the oil.ovm binary may cause this error.even
See https://github.com/oilshell/oil/issues/47
```

This message shows up after the script is exited right upon typing the next character into the shell! Strange..

**Comment by mil** [comment]
Actually, that workaround doesn't really work.. but raises another issue (likely related):

```sh
#!/usr/bin/env osh

trap handler HUP

handler() {
  echo works
}

echo before
read var & wait
echo after
```

- **bash**: returns instantly
- **ash**: returns instantly
- **oil**: waits indefinitely for read?

**Comment by andychu** [comment]
Oh actually I thought about this this morning :)  What I think we should do is make a new function `pyos.Read()` that essentially does:


```
def Read(...)
  try:
    c = posix.read(1)
  except IOError as e:
    if e.errno == errno.EINTR:
      return EINTR_VALUE
    raise
```

And then change all the callers to do `pyos.Read()` instead of `posix.read()`.

This is motivated by details of the C++ translation  -- we can avoid using C++ exceptions  by just writing our own `pyos.Read()` function in C++.  If that doesn't make sense ping me on https://oilshell.zulipchat.com :)

This is somewhat documented at the end of: https://github.com/oilshell/oil/blob/master/README.md

Unlike most code, the `pyos.py` files aren't translated directly, we instead reimplement them by hand, which allows us to avoid the exception style used by Python's API.

**Comment by andychu** [comment]
Thanks for the other report, that's a bug too!   #1001.

That is sort of a related issue in that it relates to Python exceptions.  Let's think about this one first maybe ...

**Comment by andychu** [comment]
Just for reference, I did this in the dash source code:

```
fgrep -n 'read('  * 
```

And it came up with 3 instances:

Backticks seem to retry on EINTR:

```
    do {                                                                                                                                                                                                           
      i = read(in.fd, buf, sizeof buf);                                                                                                                                                                            
    } while (i < 0 && errno == EINTR);                                                                                                                                                                             
    TRACE(("expbackq: read returns %d\n", i));    
```

And then `preadfd` has some complicated logic that calculates the number of retries!

```
                                                                                                                                                                                                                   
    if (nr < 0) {                                                                                                                                                                                                  
        if (errno == EINTR)                                                                                                                                                                                        
            goto retry;   
```

And the `read()` builtin:

```
    for (;;) {                                                                                                                                                                                                     
        switch (read(0, &c, 1)) {                                                                                                                                                                                  
        case 1:                                                                                                                                                                                                    
            break;                                                                                                                                                                                                 
        default:                                                                                                                                                                                                   
            if (errno == EINTR && !pending_sig)                                                                                                                                                                    
                continue;                                                                                                                                                                                          
                /* fall through */                                                                                                                                                                                 
        case 0:                                                                                                                                                                                                    
            status = 1;                                                                                                                                                                                            
            goto out;  
```

So this seems laborious, but I guess the lesson is that we really do have to handle EINTR everywhere ... All the `read(` invocations in dash explicitly handle it :-/

**Comment by andychu** [comment]
Also this page on how to build and run Oil should be up to date, but I will be very interested if you have problems!   I am trying to encourage contributions but I think people get stuck at some of the build steps.  It is most predictable on Debian/Ubuntu

https://github.com/oilshell/oil/wiki/Contributing

**Comment by andychu** [comment]
Thinking about this some more, I think the signature should look like:

```
# read up to 4096 bytes from the file descriptor, and append them to buf
# returns the number of bytes read, and an error

buf = []
n, err_num = pyos.Read(fd, 4096, buf)
if err_num == errno.EINTR:
  print('do something')
```

We might also want a variant that lets you test if the character is the delimiter for `-d`

```
b, err_num = pyos.ReadByte(fd, buf)
if b == ',':
  print('comma')
```

**Comment by andychu** [comment]
Testing this out a bit, I see the difference between bash and dash.  Restating what you said:

- bash: every time you send SIGHUP, it prints "sighup recvd" immediately.  But the read is retried, and eventually succeeds with status 0
- dash: when you send SIGHUP, it prints "sighup recvd" immediately, and then the read fails with status 1

Hm I'm actually not sure how to implement the bash behavior... the dash behavior might be easier to implement.

**Comment by andychu** [comment]
OK bash has explicit handling of EINTR, and then it explicitly runs traps in the `read` builtin.   I guess I should have expected that based on the behavior!

in builtin/read.def

```
      if (retval <= 0)                                                                                                                                                                                             
        {                                                                                                                                                                                                          
          if (retval < 0 && errno == EINTR)                                                                                                                                                                        
            {                                                                                                                                                                                                      
              lastsig = LASTSIG();                                                                                                                                                                                 
              if (lastsig == 0)                                                                                                                                                                                    
                lastsig = trapped_signal_received;                                                                                                                                                                 
              run_pending_traps ();     /* because interrupt_immediately is not set */                                                                                                                             
            }                                                                                                                                                                                                      
          else                                                                                                                                                                                                     
            lastsig = 0;                                                                                                                                                                                           
          if (terminating_signal && tty_modified)                                                                                                                                                                  
            ttyrestore (&termsave);     /* fix terminal before exiting */                                                                                                                                          
          CHECK_TERMSIG;                                                                                                                                                                                           
          eof = 1;                                                                                                                                                                                                 
          break;                                                                                                                                                                                                   
        }                                                                                                                                                                                                          
           
```

**Comment by andychu** [comment]
Hm in `lib/sh/zread.c`, bash also has `zread()`, `zreadretry()`, `zreadintr(),  which are variants on the EINTR behavior ...

Also `zreadn`, `zreadc`, etc.  Geez this is a lot of behavior

**Comment by andychu** [comment]
OK I actually made a lot of progress on this ... It is kind of hairy but I think I get it.  I reverted the EINTR retry in `posix.read()`, and refactored the code to use new `pyos.Read()` and `pyos.ReadByte()` instead of `posix.read()`.

So next step is to actually run the trap handlers in between retries like bash and zsh do!  I think that should be relatively straightforward

**Comment by andychu** [comment]
OK this is fixed now.  All these cases explicitly run trap handlers on EINTR:

- read
- read -n (number of bytes)
- read -d
- mapfile

It will be out with the next release, or any testing at HEAD is appreciated.  This file can be used to demo it and is based on your examples:

https://github.com/oilshell/oil/blob/master/demo/signal-during-read.sh

Thanks for the report!

**Comment by mil** [comment]
That's great to hear its resolved! Will try and test soon off HEAD.

**Comment by mil** [comment]
Fixed as of 0.9.4, thanks!

**Comment by andychu** [comment]
Great thank you for verifying all of these!  I'm writing a release announcement now, and am grateful for all the reports and testing.

We just had our first oil-dev walkthrough which has some notes on Zulip -- it was successful so I want to do something like this more regularly. Details to come!
